### ISSUE TITLE
Wallet lock when block arrives too much code is locked

### ISSUE THREAD

**Issue by dangershony** [stratisproject/StratisBitcoinFullNode#557]
The ProcessBlock method locks the entire call to ProcessTransaction some part of the code in ProcessesTransaction should not be locked (like the vent) we need to fix this.

We can split the method of ProcessesTransaction in two parts Lock logic and none Lock logic.

One for @bokobza to confirm

**Comment by bokobza** [comment]
Like I said somewhere in my comments here https://github.com/stratisproject/StratisBitcoinFullNode/commit/61c3b92f12afd3fc2b3904af990f73781a6ede69, ProcessTransaction didn't have and doesn't need to throw an event. The event was thrown at the end of AddTransactionToWallet. In the new code, the event thrown at the end of AddTransactionToWallet is replaced by a call to a method called TransactionFoundInternal.
So in conclusion:
- the event should be removed from ProcessTransaction
- the event should be deleted as it's not used anywhere
- ProcessTransaction doesn't need to be split

**Comment by dangershony** [comment]
What about ` this.LoadKeysLookupLock();` its called both before and after the lock. if we don't split the method there will be a double lock (from ProcessBlock) this is what I was trying to avoid.

If we can remove `LoadKeysLookupLock` then no need to split the method.

**Comment by bokobza** [comment]
This has nothing to do with LoadKeys, it's about the event.
What's a double lock?
this?

lock()
{
    lock()
   {
   }
}

**Comment by bokobza** [comment]
as far as I know it's not a problem.

**Comment by dangershony** [comment]
Yeah sorry its actually recursive lock, its not a problem but why not avoid it if we can.

`LoadKeysLookupLock` is inside the ProcessTransaction method.
Initially I wanted to split it so the event and the LoadKeys will be outside of the lock of the caller (ProcessesBlock) now we might also add Save to the mix.

This is my suggestion.

ProcessesBlock
- some logic
- lock(ProcessTransactionLocked)
- if found -> LoadKeys
- if found -> Save

ProcessTransaction
- LoadKeys
- lock (ProcessTransactionLocked)
- if found -> LoadKeys
- if found -> Save

ProcessTransactionLocked
- trx logic

**Comment by Aprogiena** [comment]
Recursive locks are always bad. Avoid them. Each lock should follow a documented contract (add those contracts to comments!!). Recursive locks extend such contracts to not well defined contracts.

**Comment by Aprogiena** [comment]
As mentioned in the other issue - It might also be a good idea that we first introduce the puller to the wallet, because that will change a lot of code and only then try to refactor it.

**Comment by bokobza** [comment]
Closing as this discussion is included in https://github.com/stratisproject/StratisBitcoinFullNode/issues/1103.
